nomos-da/network/core/src/protocols/replication/behaviour.rs

```diff
@@ -156,361 +218,254 @@ where
-                if !events.is_empty() {
-                    completed = false;
-
-                    for event in events {
-                        // Intercept the events that should be processed by the handler.
-                        if let ToSwarm::NotifyHandler {
-                            peer_id,
-                            handler: NotifyHandler::One(connection_id),
-                            event:
-                                Either::Left(BehaviourEventToHandler::OutgoingMessage { message }),
-                        } = event
-                        {
-                            // Deliver the message to the appropriate peer's handler.
-                            deliver_message_to_peer(
-                                &mut all_behaviours,
-                                peer_id,
-                                connection_id,
-                                message.clone(),
-                            );
+            match self.outgoing_streams.entry(*peer_id) {
+                Entry::Occupied(mut entry) => {
+                    match entry.get_mut().take() {
+                        StreamState::Idle(stream) => {
+                            let message = messages
+                                .pop_front()
+                                .expect("At least one message is queued");
+                            self.outgoing_tasks
+                                .push(Box::pin(Self::try_write_message(*peer_id, message, stream)));
+
+                            should_wake = true;
+                        }
+                        StreamState::Busy => {
+                            // We are already sending a message on this stream,
+                            // so any pending messages need to wait
                         }
                     }
                 }
-            }
-
-            // Filter out behaviors that no longer generate events.
-            pending_behaviours.retain(|i| {
-                let mut events = vec![];
-                let behaviour = &mut all_behaviours[*i];
-                while let Poll::Ready(event) = behaviour.poll(&mut cx) {
-                    events.push(event);
+                // If there is no stream for this peer yet, try to open one, and then write the
+                // first message into it
+                Entry::Vacant(entry) => {
+                    entry.insert(StreamState::Busy);
+                    let message = messages
+                        .pop_front()
+                        .expect("At least one message is queued");
+                    let control = self.control.clone();
+                    // Avoid "coercion requires that `'1` must outlive `'static`" compilation error
+                    let peer_id = *peer_id;
+                    self.outgoing_tasks.push(Box::pin(async move {
+                        let stream = Self::try_open_stream(peer_id, control).await?;
```

Comment:
I know that we talked about having stream per request, but since replication request doesn't have response or acknowledgment - we can create one stream when incoming or outgoing connection is established.

Reply:
For sure, it will be easier to see in another PR, for now we can keep it as is.